File: cpp/curvature_generator.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <set>
#include <chrono>
#include <thread>
#include <limits.h>
#include <math.h>

using namespace std;
using namespace std::chrono;

set<int> curvatures;
set<string> circles;

/**
 * Finds the inner product of two circles
 */
double inner_product(double *c_1, double *c_2){
    return -0.5*c_1[0]*c_2[1]-0.5*c_1[1]*c_2[0]+c_1[2]*c_2[2]+c_1[3]*c_2[3];
}

/**
 * Finds the coordinates after reflecting one circle across another
 * @param base the circle to be reflected
 * @param refl the circle reflected across
 * @param result return result via pointer
 */
void reflect(double *base, double *refl, double *result){
    double ip = inner_product(base, refl);
    result[0] = base[0] - 2*ip*refl[0];
    result[1] = base[1] - 2*ip*refl[1];
    result[2] = base[2] - 2*ip*refl[2];
    result[3] = base[3] - 2*ip*refl[3];
}

/**
 * Returns a string with the coordinates of the circle
 */
string circ_to_str(double *circle){
    return to_string(circle[0])+" "+to_string(circle[1])+" "+to_string(circle[2])+" "+to_string(circle[3]);
}

/**
 * Recursively finds reflections of a circle across dual circles
 * @param curvs set to store curvatures
 * @param circs vector to store reflected circles
 * @param base_circle the circle to reflect
 * @param dual_config the dual configuration
 * @param dual_size the number of circles in the dual configuration
 * @param rec_lim the maximum depth of recursions
 * @param max_num the maximum curvature to search to
 * @param scale1 scale for inversive coordinates
 * @param scale2 scale for inversive coordinates
 * @param scale3 scale for inversive coordinates
 * @param scale4 scale for inversive coordinates
 * @param print_circ option to save the circles
 */
void find_reflections_rec(set<int> *curvs, set<string> *circs, double *base_circle, double **dual_config, int dual_size, int rec_lim, int max_num, double scale1, double scale2, double scale3, double scale4, bool print_circ){

    curvs->insert(round(base_circle[1] / scale2));
    if (print_circ){
        circs->insert(circ_to_str(base_circle));
    }

    if (rec_lim == 0){
        if (max_num < INT_MAX){
            cout << "Error: recursion limit reached" << endl;
        }
        return;
    }

    for (int i=0;i<dual_size;i++){
        if (inner_product(base_circle,dual_config[i]) < -0.01){
            double *ref = new double[4];
            reflect(base_circle, dual_config[i], ref);
            //avoid precision issues
            // ref[0] = round(ref[0] / scale1 * 10) * scale1 / 10;
            // ref[1] = round(ref[1] / scale2 * 10) * scale2 / 10;
            // ref[2] = round(ref[2] / scale3 * 10) * scale3 / 10;
            // ref[3] = round(ref[3] / scale4 * 10) * scale4 / 10;
            if (ref[1] <= max_num){
                find_reflections_rec(curvs, circs, ref, dual_config, dual_size, rec_lim-1, max_num, scale1, scale2, scale3, scale4, print_circ);
            }
            delete [] ref;
        }
    }

}

/**
 * Thread handler
 */
void thread_func(set<int> *curvs, set<string> *circs, double *base_circle, double **dual_config, int dual_size, int rec_lim, int max_num, double scale1, double scale2, double scale3, double scale4, bool print_circ){
    find_reflections_rec(curvs, circs, base_circle, dual_config, dual_size, rec_lim, max_num, scale1, scale2, scale3, scale4, print_circ);
}

/**
 * Main function to handle finding reflections,
 * spawns off threads to do calculations and organises the results
 * @param base_config the base configuration
 * @param base_size the number of circles in the base configuration
 * @param dual_config the dual configuration
 * @param dual_size the number of circles in the dual configuration
 * @param rec_lim the maximum depth of recursions
 * @param max_num the maximum curvature to search to
 * @param scale1 scale for inversive coordinates
 * @param scale2 scale for inversive coordinates
 * @param scale3 scale for inversive coordinates
 * @param scale4 scale for inversive coordinates
 * @param thread_num the number of threads (currently defaults to the base size)
 * @param print_circ option to save the circles
 */
void find_reflections(double **base_config, int base_size, double **dual_config, int dual_size, int rec_lim, int max_num, double scale1, double scale2, double scale3, double scale4, int thread_num, bool print_circ){
    
    thread **threads = new thread *[thread_num];
    set<int> *curvs = new set<int>[thread_num];
    set<string> *circs = new set<string>[thread_num];
    
    int load_num = ceil((double) thread_num / 8);
    int s = 0;
    int t = 0;

    for (int i=0;i<load_num;i++){
        for (int j=0;j<8;j++){
            threads[s] = new thread(thread_func, &curvs[s], &circs[s], base_config[s], dual_config, dual_size, rec_lim, max_num, scale1, scale2, scale3, scale4, print_circ);
            s++;
            if (s >= thread_num){
                break;
            }
        }

        for (int j=0;j<8;j++){
            threads[t]->join();
            delete threads[t];
            curvatures.insert(curvs[t].begin(), curvs[t].end());
            circles.insert(circs[t].begin(), circs[t].end());
            t++;
            if (t >= thread_num){
                break;
            }
        }
    }
    
    
    delete [] threads;
    delete [] curvs;
    delete [] circs;

}

string curvs_mod(int n){
    set<int> curvs_mod;
    for (auto i : curvatures){
        int temp = i;
        while (temp < 0){
            temp += n;
        }
        curvs_mod.insert(temp%n);
    }
    string ans = "";
    for (auto i : curvs_mod){
        ans = ans + to_string(i) + " ";
    }
    ans += "\nMissing: ";
    for (int i=0;i<n;i++){
        auto j = curvs_mod.find(i);
        if (j == curvs_mod.end()){
            ans = ans + to_string(i)+ " ";
        }
    }
    return ans;
}

int main(int argc, char *argv[]){ 

    if (argc != 3){
        cout <<"Input error: please use the following format:"<< endl;
        cout <<"./curv_gen <type> <index>"<< endl;
        return 1;
    }

    //read in data

    string type = argv[1];
    string index = argv[2];

    ifstream f_base("../configs/"+type+"_"+index+"_base.txt");
    ifstream f_dual("../configs/"+type+"_"+index+"_dual.txt");

    if (!f_base.is_open() || !f_dual.is_open()){
        cout <<"Error: File cannot be opened."<< endl;
        return 1;
    }

    string in_str;
    int base_size, dual_size;

    f_base >> in_str;
    base_size = stoi(in_str);
    double **base_config = new double *[base_size];
    for (int i=0;i<base_size;i++){
        base_config[i] = new double[4];
        for (int j=0;j<4;j++){
            f_base >> in_str;
            if (in_str.find("r") != std::string::npos){
                double rat = stod(in_str);
                int rad = stoi(in_str.substr(in_str.find("r")+1));
                base_config[i][j] = rat*sqrt(rad);
            }
            else{
                base_config[i][j] = stod(in_str);
            }
        }
    }
    
    f_dual >> in_str;
    dual_size = stoi(in_str);
    double **dual_config = new double *[dual_size];
    for (int i=0;i<dual_size;i++){
        dual_config[i] = new double[4];
        for (int j=0;j<4;j++){
            f_dual >> in_str;
            if (in_str.find("r") != std::string::npos){
                double rat = stod(in_str);
                int rad = stoi(in_str.substr(in_str.find("r")+1));
                dual_config[i][j] = rat*sqrt(rad);
            }
            else{
                dual_config[i][j] = stod(in_str);
            }
        }
    }

    f_base.close();
    f_dual.close();

    double scale1, scale2, scale3, scale4;
    if (!type.compare("cube")){
        scale1 = sqrt(2);
        scale2 = sqrt(2);
        scale3 = 1;
        scale4 = sqrt(2);
    }
    else if (!type.compare("oct")){
        scale1 = 1;
        scale2 = 1;
        scale3 = sqrt(2);
        scale4 = 1;
    }
    else if (!type.compare("cubeoct")){
        scale1 = 1;
        scale2 = 1;
        scale3 = sqrt(6);
        scale4 = 1;
    }
    else if (!type.compare("square")){
        scale1 = 1;
        scale2 = 1;
        scale3 = 1;
        scale4 = 1;
    }
    else if (!type.compare("triangle")){
        scale1 = 1;
        scale2 = 1;
        scale3 = 1;
        scale4 = sqrt(3);
    }
    else if (!type.compare("hexagon")){
        scale1 = sqrt(3);
        scale2 = sqrt(3);
        scale3 = sqrt(3);
        scale4 = 1;
    }
    else{
        cout <<"Type unrecognised; enter scale factors manually:"<<endl;
        cout <<"b bar scale";
        cin >> scale1;
        cout <<"b scale";
        cin >> scale2;
        cout <<"h_1 scale";
        cin >> scale3;
        cout <<"h_2 scale";
        cin >> scale4;
    }

    int rec_lim = 200;
    int max_num = INT_MAX;

    cout <<"Mode selection:"<<endl;
    cout <<" - Recursion limited (r)"<<endl;
    cout <<" - Max curvature limited (c)"<<endl;
    cout <<"Enter the mode (r/c): ";
    cin >> in_str;
    if (in_str.compare("r") == 0){
        cout <<"How many iterations: ";
        cin >> rec_lim;
    }
    else if (in_str.compare("c") == 0){
        cout <<"Find curvatures up to: ";
        cin >> max_num;
    }
    else{
        cout <<"Input error"<<endl;
        for (int i=0;i<base_size;i++){
            delete [] base_config[i];
        }
        for (int i=0;i<dual_size;i++){
            delete [] dual_config[i];
        }
        delete [] base_config;
        delete [] dual_config;
        return 1;
    }

    bool print_circ = false;
    bool print_curv = false;
    bool print_mod = false;

    cout <<"Print circles? (y/n)";
    cin >> in_str;
    if (in_str.compare("y") == 0){
        print_circ = true;
    }
    cout <<"Print curvatures? (y/n)";
    cin >> in_str;
    if (in_str.compare("y") == 0){
        print_curv = true;
    }
    cout <<"Print modulo n? (y/n)";
    cin >> in_str;
    if (in_str.compare("y") == 0){
        print_mod = true;
    }
    
    //calculations

    auto start = high_resolution_clock::now();
    
    find_reflections(base_config, base_size, dual_config, dual_size, rec_lim, max_num, scale1, scale2, scale3, scale4, base_size, print_circ);

    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << "Calculation time: " << duration.count() * 0.000001 << " s" << endl;

    //output

    start = high_resolution_clock::now();

    if (print_curv){
        ofstream f_curv("../packings/"+type+"_"+index+"_curvatures.txt");
        for (auto i : curvatures){
            f_curv <<i<<endl;
        }
        f_curv.close();
    }

    if (print_circ){
        ofstream f_circ("circle_packing.txt");
        for (auto i : circles){
            f_circ << i <<endl;
        }
        f_circ << "END" <<endl;
        f_circ.close();
        ofstream f_circ_d("circle_packing_dual.txt");
        for (int i=0;i<dual_size;i++){
            f_circ_d <<dual_config[i][0]<<" "<<dual_config[i][1]<<" "<<dual_config[i][2]<<" "<<dual_config[i][3]<<" "<<endl;
        }
        f_circ_d << "END" <<endl;
        f_circ_d.close();
    }

    if (print_mod){
        ofstream f_mod("curvatures_mod.txt");
        f_mod <<"mod 2: "<<curvs_mod(2)<<endl<<endl;
        f_mod <<"mod 3: "<<curvs_mod(3)<<endl<<endl;
        f_mod <<"mod 4: "<<curvs_mod(4)<<endl<<endl;
        f_mod <<"mod 5: "<<curvs_mod(5)<<endl<<endl;
        f_mod <<"mod 6: "<<curvs_mod(6)<<endl<<endl;
        f_mod <<"mod 7: "<<curvs_mod(7)<<endl<<endl;
        f_mod <<"mod 8: "<<curvs_mod(8)<<endl<<endl;
        f_mod <<"mod 9: "<<curvs_mod(9)<<endl<<endl;
        f_mod <<"mod 11: "<<curvs_mod(11)<<endl<<endl;
        f_mod <<"mod 12: "<<curvs_mod(12)<<endl<<endl;
        f_mod <<"mod 13: "<<curvs_mod(13)<<endl<<endl;
        f_mod <<"mod 16: "<<curvs_mod(16)<<endl<<endl;
        f_mod <<"mod 24: "<<curvs_mod(24)<<endl<<endl;
        f_mod <<"mod 32: "<<curvs_mod(32)<<endl<<endl;
        f_mod <<"mod 48: "<<curvs_mod(48)<<endl<<endl;
    }

    stop = high_resolution_clock::now();
    duration = duration_cast<microseconds>(stop - start);
    cout << "Write time: " << duration.count() * 0.000001 << " s" << endl;

    //clean up

    for (int i=0;i<base_size;i++){
        delete [] base_config[i];
    }
    for (int i=0;i<dual_size;i++){
        delete [] dual_config[i];
    }
    delete [] base_config;
    delete [] dual_config;

    return 0;
}

File: curve_tree/curvature_generator.py
import sys
import math
import numpy as np
import timeit


class Node:
    def __init__(self,data):
        self.children = []
        self.data = data
    def __str__(self):
        return f"{self.data}; {[n.data for n in self.children]}"


with open(sys.argv[1]) as f:
    width, height = [x for x in next(f).split()]
    base = [[x for x in line.split()] for line in f]

for i in range(len(base)):
    for j in range(len(base[0])):
        if base[i][j].find('r') == -1:
            base[i][j] = float(base[i][j])
        else:
            rat = base[i][j][:base[i][j].find('r')]
            rad = base[i][j][base[i][j].find('r')+1:]
            base[i][j] = float(rat)*math.sqrt(int(rad))

with open(sys.argv[2]) as f:
    width, height = [x for x in next(f).split()]
    refl = [[x for x in line.split()] for line in f]

for i in range(len(refl)):
    for j in range(len(refl[0])):
        if refl[i][j].find('r') == -1:
            refl[i][j] = float(refl[i][j])
        else:
            rat = refl[i][j][:refl[i][j].find('r')]
            rad = refl[i][j][refl[i][j].find('r')+1:]
            refl[i][j] = float(rat)*math.sqrt(int(rad))


root_r = Node([1,0,0,0])
for i in range(len(base)):
    root_r.children.append(Node(base[i]))

def makechildren(root):
    for i in range(len(refl)):
        root.children.append(Node(i))

def reflect(c_1,c_2):
    inner_product = np.array(c_1) @ np.array([[0,-0.5,0,0],[-0.5,0,0,0],[0,0,1,0],[0,0,0,1]]) @ np.array(c_2).transpose()
    result = np.array(c_1)-2*inner_product*np.array(c_2)
    return result

def comp_lists(l_1,l_2,length):
    for i in range(length):
        if l_1[i] != l_2[i]:
            return False
    return True


curvatures = []
circles = []

limit = 1
scale = sys.argv[3]
if scale.find('r') == -1:
    scale = float(scale)
else:
    rat = scale[:scale.find('r')]
    rad = scale[scale.find('r')+1:]
    scale = float(rat)*math.sqrt(int(rad))

def find_reflections(root,level):
    if not comp_lists(root.data,[1,0,0,0],4) and not comp_lists(root.data,[0,0,0,0],4):
        curvatures.append(round(root.data[1]/scale))
        circles.append(root.data)
    if level > limit:
        return
    elif comp_lists(root.data,[0,0,0,0],4):
        return
    elif comp_lists(root.data,[1,0,0,0],4):
        for i in range(len(base)):
            find_reflections(root.children[i],level+1)
    else:
        makechildren(root)
        for i in range(len(refl)):
            root.children[i].data = reflect(root.data,refl[i])
            if comp_lists(root.children[i].data,root.data,4):
                root.children[i] = root
            find_reflections(root.children[i],level+1)


def list_mod(l,n):
    l1 = [x % n for x in l]
    l2 = list(set(l1))
    return l2

def legendre(a,p):
    #if p is prime:
    result = pow(a,((p-1)/2)) % p
    #if not we need to do something else
    #TODO: something else
    return result

start = timeit.default_timer()

find_reflections(root_r,0)

curvatures = list(set(curvatures))
curvatures.sort()

out_c = open('curvature.txt','w')
for i in range(len(curvatures)):
    out_c.write(str(curvatures[i])+'\n')
out_c.close()

out_p = open('circle_packing.txt','w')
for i in range(len(circles)):
    out_p.write(str(round(circles[i][0]*100000000)/100000000)+' ')
    out_p.write(str(round(circles[i][1]*100000000)/100000000)+' ')
    out_p.write(str(round(circles[i][2]*100000000)/100000000)+' ')
    out_p.write(str(round(circles[i][3]*100000000)/100000000)+'\n')
out_p.write('END')
out_p.close()


# print(curvatures)

# print('mod 2: ',list_mod(curvatures,2))
# print('mod 3: ',list_mod(curvatures,3))
# print('mod 4: ',list_mod(curvatures,4))
# print('mod 5: ',list_mod(curvatures,5))
# print('mod 7: ',list_mod(curvatures,7))
# print('mod 8: ',list_mod(curvatures,8))
# print('mod 9: ',list_mod(curvatures,9))
# print('mod 11: ',list_mod(curvatures,11))
# print('mod 13: ',list_mod(curvatures,13))
# print('mod 16: ',list_mod(curvatures,16))
# print('mod 24: ',list_mod(curvatures,24))
# print('mod 25: ',list_mod(curvatures,25))
# print('mod 32: ',list_mod(curvatures,32))


tangent_circles = []

def tan_circ_func_rec(root_1,root_2):
    if comp_lists(root_1.data,[1,0,0,0],4) or comp_lists(root_1.data,[0,0,0,0],4) or comp_lists(root_2.data,[1,0,0,0],4) or comp_lists(root_2.data,[0,0,0,0],4):
        return
    a = round(root_1.data[1]/scale)
    b = round(root_2.data[1]/scale)
    c = [a,b]
    c.sort()
    tangent_circles.append(str(c[0])+','+str(c[1]))
    if not root_1.children or not root_2.children:
        return
    else:
        for i in range(len(refl)):
            if not (root_1 == root_1.children[i] and root_2 == root_2.children[i]):
                tan_circ_func_rec(root_1.children[i],root_2.children[i])


def tan_circ_func(c_1,c_2):
    tan_circ_func_rec(root_r.children[c_1],root_r.children[c_2])

#oct: 0-2,0-3,0-4,0-5,1-2,1-3,1-4,1-5,2-3,2-4,3-5,4-5
if len(base) == 6:
    tan_circ_func(0,2)
    tan_circ_func(0,3)
    tan_circ_func(0,4)
    tan_circ_func(0,5)
    tan_circ_func(1,2)
    tan_circ_func(1,3)
    tan_circ_func(1,4)
    tan_circ_func(1,5)
    tan_circ_func(2,3)
    tan_circ_func(2,4)
    tan_circ_func(3,5)
    tan_circ_func(4,5)
#cube: 0-1,0-2,0-4,1-3,1-5,2-3,2-6,3-7,4-5,4-6,5-7,6-7
elif len(base) == 8:
    tan_circ_func(0,1)
    tan_circ_func(0,2)
    tan_circ_func(0,4)
    tan_circ_func(1,3)
    tan_circ_func(1,5)
    tan_circ_func(2,3)
    tan_circ_func(2,6)
    tan_circ_func(3,7)
    tan_circ_func(4,5)
    tan_circ_func(4,6)
    tan_circ_func(5,7)
    tan_circ_func(6,7)


tangent_circles = list(set(tangent_circles))
tangent_circles.sort()

tangent_circles_1 = [[int(x[:x.find(',')]),int(x[x.find(',')+1:])] for x in tangent_circles]

out_t = open('tangent_circles.txt','w')
for i in range(len(tangent_circles_1)):
    out_t.write(str(tangent_circles_1[i][0])+' '+str(tangent_circles_1[i][1])+'\n')
    out_t.write(str(tangent_circles_1[i][1])+' '+str(tangent_circles_1[i][0])+'\n')
out_t.close()

#print(tangent_circles_1)

tuples = []

def tuple_func_rec(roots,amt):
    for i in range(amt):
        if comp_lists(roots[i].data,[1,0,0,0],4):
            return
    a = []
    for i in range(amt):
        a.append(round(roots[i].data[1]/scale))
    b = ''
    for i in range(amt-1):
        b += str(a[i])+','
    b += str(a[amt-1])
    tuples.append(b)
    for i in range(amt):
        if not roots[i].children:
            return
    for j in range(len(refl)):
        for i in range(amt):
            if roots[i] == roots[i].children[j]:
                continue
        new_roots = []
        for i in range(amt):
            new_roots.append(roots[i].children[j])
        tuple_func_rec(new_roots,amt)

tuple_func_rec(root_r.children,6)

tuples = list(set(tuples))
tuples.sort()

out_t = open('tuples.txt','w')
for i in range(len(tuples)):
    out_t.write(tuples[i]+'\n')
out_t.close()


stop = timeit.default_timer()
print('Time: ', stop - start)